lib/install/plugin/register.go

```diff
@@ -0,0 +1,212 @@
+	Context          context.Context
+
+	info *Info
+
+	tURL      *url.URL
+	connected bool
+}
+
+func NewPluginer(ctx context.Context, target *url.URL, i *Info) (*Pluginer, error) {
+	defer trace.End(trace.Begin(""))
+
+	p := &Pluginer{
+		tURL: target,
+		info: i,
+	}
+	p.Context = ctx
+
+	err := p.connect()
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+func (p *Pluginer) disconnect() error {
+	defer trace.End(trace.Begin(""))
+	if p.Session != nil {
+		err := p.Session.Client.Logout(p.Context)
+		if err != nil {
+			return fmt.Errorf("failed to disconnect: %s", err)
+		}
+	}
+	p.connected = false
+	return nil
+}
+
+func (p *Pluginer) connect() error {
+	defer trace.End(trace.Begin(""))
+	var err error
+
+	sessionconfig := &session.Config{
+		Insecure: true,
+	}
+	sessionconfig.Service = p.tURL.String()
+
+	p.Session = session.NewSession(sessionconfig)
+	p.Session, err = p.Session.Connect(p.Context)
+	if err != nil {
+		return fmt.Errorf("failed to connect: %s", err)
+	}
+
+	p.Session.Populate(p.Context)
+
+	em, err := object.GetExtensionManager(p.Session.Client.Client)
+	if err != nil {
+		return fmt.Errorf("failed to get extension manager: %s", err)
+	}
+	p.ExtensionManager = em
+
+	p.connected = true
+	return nil
+}
+
+// Register installs an extension to the target
+func (p *Pluginer) Register() error {
+	defer trace.End(trace.Begin(""))
+	var err error
+	if !p.connected {
+		return errors.New("not connected")
+	}
+
+	desc := types.Description{
+		Label:   p.info.Name,
+		Summary: p.info.Summary,
+	}
+
+	e := types.Extension{
+		Key:         p.info.Key,
+		Version:     p.info.Version,
+		Company:     p.info.Company,
+		Description: &desc,
+	}
+
+	eci := types.ExtensionClientInfo{
+		Version:     p.info.Version,
+		Company:     p.info.Company,
+		Description: &desc,
+		Type:        p.info.Type,
+		Url:         p.info.URL,
+	}
+	e.Client = append(e.Client, eci)
+
+	d := types.KeyValue{
+		Key:   "name",
+		Value: p.info.Name,
+	}
+
+	eri := types.ExtensionResourceInfo{
+		Locale: "en_US",
+		Module: "name",
+	}
+
+	eri.Data = append(eri.Data, d)
+
+	e.ResourceList = append(e.ResourceList, eri)
+
+	// TODO HTTP server info
+
+	e.ShownInSolutionManager = &p.info.ShowInSolutionManager
+
+	e.LastHeartbeatTime = time.Now().UTC()
+
+	err = p.ExtensionManager.Register(p.Context, e)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Unregister removes an extension from the target
+func (p *Pluginer) Unregister(key string) error {
+	defer trace.End(trace.Begin(""))
+	if !p.connected {
+		return errors.New("not connected")
+	}
+
+	if err := p.ExtensionManager.Unregister(p.Context, key); err != nil {
+		return err
+	}
+	return nil
+}
+
+// IsRegistered checks for presence of an extension on the target
+func (p *Pluginer) IsRegistered(key string) (bool, error) {
+	defer trace.End(trace.Begin(""))
+	if !p.connected {
+		return false, errors.New("not connected")
+	}
+
+	l, err := p.Extensions()
```

Comment:
extensionManager has method to find one extension with key, so use that method, instead of list everything and filter by yourself.